src/query/sql/src/planner/optimizer/rule/rewrite/rule_filter_nulls.rs

```diff
@@ -0,0 +1,191 @@
+// Copyright 2021 Datafuse Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use databend_common_exception::Result;
+
+use crate::optimizer::extract::Matcher;
+use crate::optimizer::rule::Rule;
+use crate::optimizer::rule::TransformResult;
+use crate::optimizer::RelExpr;
+use crate::optimizer::RuleID;
+use crate::optimizer::SExpr;
+use crate::optimizer::Statistics;
+use crate::plans::Filter;
+use crate::plans::FunctionCall;
+use crate::plans::Join;
+use crate::plans::JoinType;
+use crate::plans::RelOp;
+use crate::plans::RelOperator;
+use crate::ScalarExpr;
+
+const NULL_THRESHOLD_RATIO: f64 = 0.02;
```

Comment:
I think this threshold should be increased, perhaps 0.5 is a good value. The reasons are: (1) DataBlock will be re-generated during the filter, which will be expensive if there are many columns; (2) During the hash join probe phase, keys that are null will not probe the hash table.

Reply:
Sure, but the 0.5 seems a little large. Prunning even 10% of the parts will give good results, especially in distributed scenarios, and reduce network overhead. But I don't have a clear value for how much.

Reply:
Since most of the cases are standalone mode and the probe and build of join will naturally filter out null values ​​(other operators will also do this), the is_not_null filter will cause `DataBlock::take` overhead, so it would be better to enable this optimization only in distributed mode. We can use 0.2, which is also used in FilterExecutor.

Reply:
Let me do something performance tests